File: src/lib/api/reddit.ts
import { SocialPost } from "@/types/social";

const REDDIT_CLIENT_ID = process.env.REDDIT_CLIENT_ID ?? "";
const REDDIT_CLIENT_SECRET = process.env.REDDIT_CLIENT_SECRET ?? "";
const REDDIT_USER_AGENT = process.env.REDDIT_USER_AGENT ?? "TechFin/1.0";

interface TokenCache {
  token: string;
  expiresAt: number;
}

let tokenCache: TokenCache | null = null;

async function getRedditToken(): Promise<string> {
  if (tokenCache && Date.now() < tokenCache.expiresAt - 60_000) {
    return tokenCache.token;
  }

  const credentials = Buffer.from(`${REDDIT_CLIENT_ID}:${REDDIT_CLIENT_SECRET}`).toString("base64");
  const response = await fetch("https://www.reddit.com/api/v1/access_token", {
    method: "POST",
    headers: {
      Authorization: `Basic ${credentials}`,
      "Content-Type": "application/x-www-form-urlencoded",
      "User-Agent": REDDIT_USER_AGENT,
    },
    body: "grant_type=client_credentials",
  });

  if (!response.ok) {
    throw new Error(`Reddit auth error: ${response.status}`);
  }

  const data = await response.json();
  tokenCache = {
    token: data.access_token,
    expiresAt: Date.now() + data.expires_in * 1000,
  };
  return tokenCache.token;
}

export async function fetchRedditPosts(subreddit: string): Promise<SocialPost[]> {
  const hasCredentials =
    REDDIT_CLIENT_ID &&
    REDDIT_CLIENT_ID !== "your_reddit_client_id" &&
    REDDIT_CLIENT_SECRET &&
    REDDIT_CLIENT_SECRET !== "your_reddit_client_secret";

  if (!hasCredentials) {
    console.warn("[reddit] No credentials configured, returning mock data");
    return getMockRedditPosts(subreddit);
  }

  try {
    const token = await getRedditToken();
    const response = await fetch(
      `https://oauth.reddit.com/r/${subreddit}/hot?limit=10`,
      {
        headers: {
          Authorization: `Bearer ${token}`,
          "User-Agent": REDDIT_USER_AGENT,
        },
      }
    );

    if (response.status === 401) {
      tokenCache = null;
      const freshToken = await getRedditToken();
      const retry = await fetch(
        `https://oauth.reddit.com/r/${subreddit}/hot?limit=10`,
        {
          headers: {
            Authorization: `Bearer ${freshToken}`,
            "User-Agent": REDDIT_USER_AGENT,
          },
        }
      );
      if (!retry.ok) throw new Error(`Reddit API error: ${retry.status}`);
      const data = await retry.json();
      return parseRedditData(data, subreddit);
    }

    if (!response.ok) {
      throw new Error(`Reddit API error: ${response.status}`);
    }

    const data = await response.json();
    return parseRedditData(data, subreddit);
  } catch (error) {
    console.error("[reddit] Fetch error:", error);
    return getMockRedditPosts(subreddit);
  }
}

function parseRedditData(data: Record<string, unknown>, subreddit: string): SocialPost[] {
  const posts = (data?.data as Record<string, unknown>)?.children as Array<Record<string, unknown>>;
  if (!Array.isArray(posts)) return getMockRedditPosts(subreddit);

  return posts.map((post) => {
    const p = post.data as Record<string, unknown>;
    return {
      id: p.id as string,
      platform: "reddit" as const,
      author: `u/${p.author as string}`,
      content: (p.selftext as string)?.slice(0, 500) || (p.title as string),
      upvotes: p.score as number,
      comments: p.num_comments as number,
      subreddit: `r/${subreddit}`,
      url: `https://reddit.com${p.permalink as string}`,
      publishedAt: new Date((p.created_utc as number) * 1000).toISOString(),
    };
  });
}

function getMockRedditPosts(subreddit: string): SocialPost[] {
  const mockPosts: Record<string, SocialPost[]> = {
    wallstreetbets: [
      {
        id: "wsb-1",
        platform: "reddit",
        author: "u/RetailInvestorGuru",
        content: "NVDA calls printing 🚀 AI momentum is unstoppable. Everyone sleeping on the GPU shortage thesis.",
        upvotes: 4823,
        comments: 342,
        subreddit: "r/wallstreetbets",
        url: "#",
        publishedAt: new Date(Date.now() - 1000 * 60 * 45).toISOString(),
      },
      {
        id: "wsb-2",
        platform: "reddit",
        author: "u/TendieHunter2000",
        content: "SPY puts. I have my reasons. See you all on the other side. Not financial advice.",
        upvotes: 2156,
        comments: 198,
        subreddit: "r/wallstreetbets",
        url: "#",
        publishedAt: new Date(Date.now() - 1000 * 60 * 90).toISOString(),
      },
      {
        id: "wsb-3",
        platform: "reddit",
        author: "u/DiamondHandsOnly",
        content: "TSLA to $400 EOY. Elon is building the future and y'all are too scared to see it. Loading up on shares.",
        upvotes: 1789,
        comments: 567,
        subreddit: "r/wallstreetbets",
        url: "#",
        publishedAt: new Date(Date.now() - 1000 * 60 * 120).toISOString(),
      },
    ],
    stocks: [
      {
        id: "stocks-1",
        platform: "reddit",
        author: "u/ValueInvestorPro",
        content: "Energy sector is severely undervalued relative to earnings. XOM and CVX both trading below historical P/E. Adding to positions.",
        upvotes: 892,
        comments: 134,
        subreddit: "r/stocks",
        url: "#",
        publishedAt: new Date(Date.now() - 1000 * 60 * 30).toISOString(),
      },
      {
        id: "stocks-2",
        platform: "reddit",
        author: "u/FundamentalsFocus",
        content: "Healthcare sector analysis: UNH and JNJ both showing strong cash flow metrics. The regulatory risk is overpriced by the market.",
        upvotes: 654,
        comments: 89,
        subreddit: "r/stocks",
        url: "#",
        publishedAt: new Date(Date.now() - 1000 * 60 * 60 * 2).toISOString(),
      },
      {
        id: "stocks-3",
        platform: "reddit",
        author: "u/DividendDave",
        content: "Building a dividend growth portfolio for retirement. KO, PG, JNJ, and VZ form the core. Reinvesting every quarter.",
        upvotes: 1205,
        comments: 210,
        subreddit: "r/stocks",
        url: "#",
        publishedAt: new Date(Date.now() - 1000 * 60 * 60 * 5).toISOString(),
      },
    ],
    economics: [
      {
        id: "econ-1",
        platform: "reddit",
        author: "u/MacroMindset",
        content: "CPI came in at 3.1% YoY. Still above Fed target but trajectory is clearly downward. Rate cuts in H2 2024 seem increasingly likely.",
        upvotes: 1456,
        comments: 287,
        subreddit: "r/economics",
        url: "#",
        publishedAt: new Date(Date.now() - 1000 * 60 * 60).toISOString(),
      },
      {
        id: "econ-2",
        platform: "reddit",
        author: "u/MonetaryPolicyWatch",
        content: "The yield curve inversion has persisted for over 18 months now. Historical correlation with recession is strong. Are we in a soft landing or just delayed?",
        upvotes: 987,
        comments: 445,
        subreddit: "r/economics",
        url: "#",
        publishedAt: new Date(Date.now() - 1000 * 60 * 60 * 3).toISOString(),
      },
      {
        id: "econ-3",
        platform: "reddit",
        author: "u/GlobalTradeAnalyst",
        content: "US trade deficit widened again in Q4. Dollar strength continues to hurt exporters. Watch industrials and materials sectors for continued pressure.",
        upvotes: 743,
        comments: 156,
        subreddit: "r/economics",
        url: "#",
        publishedAt: new Date(Date.now() - 1000 * 60 * 60 * 6).toISOString(),
      },
    ],
  };

  return mockPosts[subreddit] ?? mockPosts["stocks"];
}

File: backend/app/services/reddit.py
import logging
import time
from base64 import b64encode
from datetime import datetime, timedelta, timezone

import httpx

from app.config import settings

logger = logging.getLogger(__name__)

# Module-level token cache
_token_cache: dict | None = None


async def _get_reddit_token() -> str:
    global _token_cache

    if _token_cache and time.time() < _token_cache["expires_at"] - 60:
        return _token_cache["token"]

    credentials = b64encode(
        f"{settings.REDDIT_CLIENT_ID}:{settings.REDDIT_CLIENT_SECRET}".encode()
    ).decode()

    async with httpx.AsyncClient() as client:
        response = await client.post(
            "https://www.reddit.com/api/v1/access_token",
            headers={
                "Authorization": f"Basic {credentials}",
                "Content-Type": "application/x-www-form-urlencoded",
                "User-Agent": settings.REDDIT_USER_AGENT,
            },
            content="grant_type=client_credentials",
            timeout=10.0,
        )
        response.raise_for_status()
        data = response.json()

    _token_cache = {
        "token": data["access_token"],
        "expires_at": time.time() + data["expires_in"],
    }
    return _token_cache["token"]


def _parse_reddit_data(data: dict, subreddit: str) -> list[dict]:
    children = data.get("data", {}).get("children", [])
    if not isinstance(children, list):
        return _mock_reddit_posts(subreddit)

    posts = []
    for post in children:
        p = post.get("data", {})
        content = (p.get("selftext") or "")[:500] or p.get("title", "")
        posts.append({
            "id": p.get("id", ""),
            "platform": "reddit",
            "author": f"u/{p.get('author', 'unknown')}",
            "content": content,
            "upvotes": p.get("score", 0),
            "comments": p.get("num_comments", 0),
            "subreddit": f"r/{subreddit}",
            "url": f"https://reddit.com{p.get('permalink', '')}",
            "published_at": datetime.fromtimestamp(
                p.get("created_utc", 0), tz=timezone.utc
            ).isoformat(),
        })
    return posts


def _mock_reddit_posts(subreddit: str) -> list[dict]:
    now = datetime.now(timezone.utc)
    mock_posts: dict[str, list[dict]] = {
        "wallstreetbets": [
            {
                "id": "wsb-1",
                "platform": "reddit",
                "author": "u/RetailInvestorGuru",
                "content": "NVDA calls printing \U0001f680 AI momentum is unstoppable. Everyone sleeping on the GPU shortage thesis.",
                "upvotes": 4823,
                "comments": 342,
                "subreddit": "r/wallstreetbets",
                "url": "#",
                "published_at": (now - timedelta(minutes=45)).isoformat(),
            },
            {
                "id": "wsb-2",
                "platform": "reddit",
                "author": "u/TendieHunter2000",
                "content": "SPY puts. I have my reasons. See you all on the other side. Not financial advice.",
                "upvotes": 2156,
                "comments": 198,
                "subreddit": "r/wallstreetbets",
                "url": "#",
                "published_at": (now - timedelta(minutes=90)).isoformat(),
            },
            {
                "id": "wsb-3",
                "platform": "reddit",
                "author": "u/DiamondHandsOnly",
                "content": "TSLA to $400 EOY. Elon is building the future and y'all are too scared to see it. Loading up on shares.",
                "upvotes": 1789,
                "comments": 567,
                "subreddit": "r/wallstreetbets",
                "url": "#",
                "published_at": (now - timedelta(minutes=120)).isoformat(),
            },
        ],
        "stocks": [
            {
                "id": "stocks-1",
                "platform": "reddit",
                "author": "u/ValueInvestorPro",
                "content": "Energy sector is severely undervalued relative to earnings. XOM and CVX both trading below historical P/E. Adding to positions.",
                "upvotes": 892,
                "comments": 134,
                "subreddit": "r/stocks",
                "url": "#",
                "published_at": (now - timedelta(minutes=30)).isoformat(),
            },
            {
                "id": "stocks-2",
                "platform": "reddit",
                "author": "u/FundamentalsFocus",
                "content": "Healthcare sector analysis: UNH and JNJ both showing strong cash flow metrics. The regulatory risk is overpriced by the market.",
                "upvotes": 654,
                "comments": 89,
                "subreddit": "r/stocks",
                "url": "#",
                "published_at": (now - timedelta(hours=2)).isoformat(),
            },
            {
                "id": "stocks-3",
                "platform": "reddit",
                "author": "u/DividendDave",
                "content": "Building a dividend growth portfolio for retirement. KO, PG, JNJ, and VZ form the core. Reinvesting every quarter.",
                "upvotes": 1205,
                "comments": 210,
                "subreddit": "r/stocks",
                "url": "#",
                "published_at": (now - timedelta(hours=5)).isoformat(),
            },
        ],
        "economics": [
            {
                "id": "econ-1",
                "platform": "reddit",
                "author": "u/MacroMindset",
                "content": "CPI came in at 3.1% YoY. Still above Fed target but trajectory is clearly downward. Rate cuts in H2 2024 seem increasingly likely.",
                "upvotes": 1456,
                "comments": 287,
                "subreddit": "r/economics",
                "url": "#",
                "published_at": (now - timedelta(hours=1)).isoformat(),
            },
            {
                "id": "econ-2",
                "platform": "reddit",
                "author": "u/MonetaryPolicyWatch",
                "content": "The yield curve inversion has persisted for over 18 months now. Historical correlation with recession is strong. Are we in a soft landing or just delayed?",
                "upvotes": 987,
                "comments": 445,
                "subreddit": "r/economics",
                "url": "#",
                "published_at": (now - timedelta(hours=3)).isoformat(),
            },
            {
                "id": "econ-3",
                "platform": "reddit",
                "author": "u/GlobalTradeAnalyst",
                "content": "US trade deficit widened again in Q4. Dollar strength continues to hurt exporters. Watch industrials and materials sectors for continued pressure.",
                "upvotes": 743,
                "comments": 156,
                "subreddit": "r/economics",
                "url": "#",
                "published_at": (now - timedelta(hours=6)).isoformat(),
            },
        ],
    }
    return mock_posts.get(subreddit, mock_posts["stocks"])


async def fetch_reddit_posts(subreddit: str) -> list[dict]:
    global _token_cache

    has_credentials = (
        settings.REDDIT_CLIENT_ID
        and settings.REDDIT_CLIENT_ID != "your_reddit_client_id"
        and settings.REDDIT_CLIENT_SECRET
        and settings.REDDIT_CLIENT_SECRET != "your_reddit_client_secret"
    )

    if not has_credentials:
        logger.warning("[reddit] No credentials configured, returning mock data")
        return _mock_reddit_posts(subreddit)

    try:
        token = await _get_reddit_token()

        async with httpx.AsyncClient() as client:
            response = await client.get(
                f"https://oauth.reddit.com/r/{subreddit}/hot",
                params={"limit": "10"},
                headers={
                    "Authorization": f"Bearer {token}",
                    "User-Agent": settings.REDDIT_USER_AGENT,
                },
                timeout=10.0,
            )

            if response.status_code == 401:
                _token_cache = None
                token = await _get_reddit_token()
                response = await client.get(
                    f"https://oauth.reddit.com/r/{subreddit}/hot",
                    params={"limit": "10"},
                    headers={
                        "Authorization": f"Bearer {token}",
                        "User-Agent": settings.REDDIT_USER_AGENT,
                    },
                    timeout=10.0,
                )
                response.raise_for_status()

            elif not response.is_success:
                response.raise_for_status()

            return _parse_reddit_data(response.json(), subreddit)

    except Exception as e:
        logger.error("[reddit] Fetch error: %s", e)
        return _mock_reddit_posts(subreddit)
